<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>If I understand correctly, coloc.abf() can be run with correlated variants, that is, no prerequisite for taking through LD pruning/clumping is required. Am I correct in my understanding ? • coloc</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png"><script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="If I understand correctly, coloc.abf() can be run with correlated variants, that is, no prerequisite for taking through LD pruning/clumping is required. Am I correct in my understanding ?"><meta property="og:image" content="https://chr1swallace.github.io/coloc/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">coloc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">6.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="articles/a01_intro.html">Coloc: a package for colocalisation analyses</a></li>
    <li><a class="dropdown-item" href="articles/a02_data.html">Coloc: data structures</a></li>
    <li><a class="dropdown-item" href="articles/a03_enumeration.html">Coloc: under a single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a></li>
    <li><a class="dropdown-item" href="articles/a06_SuSiE.html">Coloc: using SuSiE to relax the single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="articles/a07_variant-specific-priors.html">Coloc: using variant-specific priros in coloc</a></li>
    <li><a class="dropdown-item" href="articles/a05_conditioning.html">DEPRECATED Coloc: relaxing the single causal variant assumption</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chr1swallace/coloc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-title-body">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="logo.png" class="logo" alt=""><h1>If I understand correctly, coloc.abf() can be run with correlated variants, that is, no prerequisite for taking through LD pruning/clumping is required. Am I correct in my understanding ?</h1>
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/HEAD/FAQ.md" class="external-link"><code>FAQ.md</code></a></small>
    </div>


<p><strong>Table of Contents</strong></p>
<ul><li><a href="#if-i-understand-correctly-colocabf-can-be-run-with-correlated-variants-that-is-no-prerequisite-for-taking-through-ld-pruningclumping-is-required-am-i-correct-in-my-understanding-">If I understand correctly, coloc.abf() can be run with correlated variants, that is, no prerequisite for taking through LD pruning/clumping is required. Am I correct in my understanding ?</a></li>
<li><a href="#assume-i-identify-a-sentinel-variant-for-a-block-of-genome-can-i-do-a-comparison-with-just-one-variant-using-colocabf">Assume I identify a sentinel variant for a block of genome, can I do a comparison with just one variant using coloc.abf()?</a></li>
<li><a href="#can-the-process-of-identifying-colocalized-variants-be-carried-out-genome-wide">Can the process of identifying colocalized variants be carried out genome wide?</a></li>
<li><a href="#how-is-colocabf-accounting-for-the-correlated-variants">How is coloc.abf accounting for the correlated variants?</a></li>
<li><a href="#how-to-define-priors-is-it-dependent-on-sample-size-or-any-other-parameters">How to define priors, is it dependent on sample size or any other parameters?</a></li>
<li><a href="#what-does-high-pp4-mean">What does high PP4 mean?</a></li>
</ul><!-- markdown-toc end --><div id="if-i-understand-correctly-colocabf-can-be-run-with-correlated-variants-that-is-no-prerequisite-for-taking-through-ld-pruningclumping-is-required-am-i-correct-in-my-understanding-" class="section level1">

<p>Yes, coloc.abf() and coloc.signals() assume they are given a dense map of all SNPs in a region that could be causal. Do not prune and clump.</p>
</div>
<div class="section level1">
<h1 id="assume-i-identify-a-sentinel-variant-for-a-block-of-genome-can-i-do-a-comparison-with-just-one-variant-using-colocabf">Assume I identify a sentinel variant for a block of genome, can I do a comparison with just one variant using coloc.abf()?<a class="anchor" aria-label="anchor" href="#assume-i-identify-a-sentinel-variant-for-a-block-of-genome-can-i-do-a-comparison-with-just-one-variant-using-colocabf"></a></h1>
<p>No, coloc.abf() and coloc.signals() assume they are given a dense map of all SNPs in a region that could be causal. This means you need to give all SNPs in a region. You can imagine they ask whether the patterns “match” across this region of SNPs, and a single variant does not represent a pattern.</p>
</div>
<div class="section level1">
<h1 id="can-the-process-of-identifying-colocalized-variants-be-carried-out-genome-wide">Can the process of identifying colocalized variants be carried out genome wide?<a class="anchor" aria-label="anchor" href="#can-the-process-of-identifying-colocalized-variants-be-carried-out-genome-wide"></a></h1>
<p>Coloc is designed to address whether two traits share causal variant(s) in a genomic region. It leaves the definition of “region” up to the user. You need to break the genome into smaller regions, within which it is reasonable to assume there is at most one (coloc.abf) or a small number (coloc.signals) of causal variants per trait. One way to do this is to use the boundaries defined by recombination hotspots, proxied by <a href="https://bitbucket.org/nygcresearch/ldetect-data/src/master/" class="external-link">this map</a> created by <a href="https://academic.oup.com/bioinformatics/article/32/2/283/1743626" class="external-link">lddetect</a>.</p>
<p>How big should a region be? Big enough that all variants in LD with a lead SNP are included; small enough that only one or a small number of causal variants might exist in it. I have found using densely genotyped studies and the lddetect boundaries above that regions typically contain 1,000-10,000 SNPs.</p>
</div>
<div class="section level1">
<h1 id="how-is-colocabf-accounting-for-the-correlated-variants">How is coloc.abf accounting for the correlated variants?<a class="anchor" aria-label="anchor" href="#how-is-colocabf-accounting-for-the-correlated-variants"></a></h1>
<p>That’s really how coloc works - by exploiting a dense SNP map - please see the <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1004383" class="external-link">original paper</a></p>
</div>
<div class="section level1">
<h1 id="how-to-define-priors-is-it-dependent-on-sample-size-or-any-other-parameters">How to define priors, is it dependent on sample size or any other parameters?<a class="anchor" aria-label="anchor" href="#how-to-define-priors-is-it-dependent-on-sample-size-or-any-other-parameters"></a></h1>
<p>This is described in detail in the <a href="https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008720" class="external-link">latest paper</a></p>
</div>
<div class="section level1">
<h1 id="what-does-high-pp4-mean">What does high PP4 mean?<a class="anchor" aria-label="anchor" href="#what-does-high-pp4-mean"></a></h1>
<p>The summary printed on the screen by coloc.abf() and coloc.signals() shows the posterior probability of whether a shared causal variant exists in the region. High PP4 does not mean all variants are causal and shared - to check which variants are most likely to be causal look at the SNP.PP column in the returned detailed results data.frame.</p>
</div>
<div class="section level1">
<h1 id="i-have-a-question-how-should-we-interpret-a-high-overall-pp4-95-but-a-low-pp4-for-single-snp-50-in-colocabf-in-this-scenario-pp0-pp3-were-small-than-00001">I have a question. How should we interpret a high overall PP4 (&gt;95%) but a low PP4 for single SNP (&lt;50%) in coloc.abf? In this scenario, PP0-PP3 were small than 0.0001.<a class="anchor" aria-label="anchor" href="#i-have-a-question-how-should-we-interpret-a-high-overall-pp4-95-but-a-low-pp4-for-single-snp-50-in-colocabf-in-this-scenario-pp0-pp3-were-small-than-00001"></a></h1>
<p>high overall PP4 (&gt;95%) means there is a high probability of colocalisation</p>
<p>low PP4 for single SNP (&lt;50%) means we cannot identify which individual SNP is jointly causal with confidence</p>
</div>


  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Wallace, Claudia Giambartolomei, Tom Willis, Jeffrey Pullin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

