<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Coloc: sensitivity to prior values • coloc</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Coloc: sensitivity to prior values">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">4.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a>
    </li>
    <li>
      <a href="../articles/a02_proportionality.html">Coloc: proportionality testing</a>
    </li>
    <li>
      <a href="../articles/a03_enumeration.html">Coloc: enumeration of configurations under a single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a>
    </li>
    <li>
      <a href="../articles/a05_conditioning.html">Coloc: relaxing the single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/simdata.html">Code to simulate a small dataset</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/chr1swallace/coloc">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Coloc: sensitivity to prior values</h1>
                        <h4 class="author">Chris Wallace</h4>
            
            <h4 class="date">2020-03-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/master/vignettes/a04_sensitivity.Rmd"><code>vignettes/a04_sensitivity.Rmd</code></a></small>
      <div class="hidden name"><code>a04_sensitivity.Rmd</code></div>

    </div>

    
    
<div id="simulate-a-small-dataset" class="section level2">
<h2 class="hasAnchor">
<a href="#simulate-a-small-dataset" class="anchor"></a>Simulate a small dataset</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(mvtnorm)
simx &lt;-<span class="st"> </span><span class="cf">function</span>(nsnps,nsamples,S,<span class="dt">maf=</span><span class="fl">0.1</span>) {
    mu &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">0</span>,nsnps)
    rawvars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/pkg/mvtnorm/man/Mvnorm.html">rmvnorm</a></span>(<span class="dt">n=</span>nsamples, <span class="dt">mean=</span>mu, <span class="dt">sigma=</span>S)
    pvars &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">pnorm</a></span>(rawvars)
    x &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Binomial.html">qbinom</a></span>(<span class="dv">1</span><span class="op">-</span>pvars, <span class="dv">2</span>, maf)
}


sim.data &lt;-<span class="st"> </span><span class="cf">function</span>(<span class="dt">nsnps=</span><span class="dv">50</span>,<span class="dt">nsamples=</span><span class="dv">200</span>,<span class="dt">causals=</span><span class="kw"><a href="https://rdrr.io/r/base/Round.html">floor</a></span>(nsnps<span class="op">/</span><span class="dv">2</span>),<span class="dt">nsim=</span><span class="dv">1</span>) {
  <span class="kw"><a href="https://rdrr.io/r/base/cat.html">cat</a></span>(<span class="st">"Generate"</span>,nsim,<span class="st">"small sets of data</span><span class="ch">\n</span><span class="st">"</span>)
  ntotal &lt;-<span class="st"> </span>nsnps <span class="op">*</span><span class="st"> </span>nsamples <span class="op">*</span><span class="st"> </span>nsim
  S &lt;-<span class="st"> </span>(<span class="dv">1</span> <span class="op">-</span><span class="st"> </span>(<span class="kw"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/outer.html">outer</a></span>(<span class="dv">1</span><span class="op">:</span>nsnps,<span class="dv">1</span><span class="op">:</span>nsnps,<span class="st">`</span><span class="dt">-</span><span class="st">`</span>))<span class="op">/</span>nsnps))<span class="op">^</span><span class="dv">4</span>
  X1 &lt;-<span class="st"> </span><span class="kw">simx</span>(nsnps,ntotal,S)
  X2 &lt;-<span class="st"> </span><span class="kw">simx</span>(nsnps,ntotal,S)
  Y1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(ntotal,<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(X1[,causals,<span class="dt">drop=</span><span class="ot">FALSE</span>]<span class="op">/</span><span class="dv">2</span>),<span class="dv">2</span>)
  Y2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(ntotal,<span class="kw"><a href="https://rdrr.io/r/base/colSums.html">rowSums</a></span>(X2[,causals,<span class="dt">drop=</span><span class="ot">FALSE</span>]<span class="op">/</span><span class="dv">2</span>),<span class="dv">2</span>)
  <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(X1) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(X2) &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste</a></span>(<span class="st">"s"</span>,<span class="dv">1</span><span class="op">:</span>nsnps,<span class="dt">sep=</span><span class="st">""</span>)
  df1 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="dt">Y=</span>Y1,X1)
  df2 &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span>(<span class="dt">Y=</span>Y2,X2)
  <span class="cf">if</span>(nsim<span class="op">==</span><span class="dv">1</span>) {
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(<span class="kw">new</span>(<span class="st">"simdata"</span>,
               <span class="dt">df1=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df1),
               <span class="dt">df2=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df2)))
  } <span class="cf">else</span> {
    index &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/split.html">split</a></span>(<span class="dv">1</span><span class="op">:</span>(nsamples <span class="op">*</span><span class="st"> </span>nsim), <span class="kw"><a href="https://rdrr.io/r/base/rep.html">rep</a></span>(<span class="dv">1</span><span class="op">:</span>nsim, nsamples))
    objects &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(index, <span class="cf">function</span>(i) <span class="kw">new</span>(<span class="st">"simdata"</span>, <span class="dt">df1=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df1[i,]),
                                             <span class="dt">df2=</span><span class="kw"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span>(df2[i,])))
    <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span>(objects)
  }
}

<span class="kw"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="dv">46411</span>)
data &lt;-<span class="st"> </span><span class="kw">sim.data</span>()</code></pre></div>
<pre><code>## Generate 1 small sets of data</code></pre>
</div>
<div id="sensitivity-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#sensitivity-analysis" class="anchor"></a>Sensitivity analysis</h1>
<p>Specifying prior values for coloc.abf() is important, as results can be dependent on these values. Defaults of <span class="math inline">\(p_1=p_2=10^{-4}\)</span> seem justified in a wide range of scenarios, because these broadly correspond to a 99% belief that there is true association when we see <span class="math inline">\(p&lt;5\times 10^{-8}\)</span> in a GWAS. However, choice of <span class="math inline">\(p_{12}\)</span> is more difficult. We hope the <a href="https://chr1swallace.shinyapps.io/coloc-priors/">coloc explorer app</a> will be helpful in exploring what various choices mean, at a per-SNP and per-hypothesis level. However, having conducted an enumeration-based coloc analysis, it is still helpful to check that any inference about colocalisation is robust to variations in prior values specified.</p>
<p>Continuing on from <a href="the%20last%20vignette">03_enumeration</a>, we have</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(coloc)</code></pre></div>
<pre><code>## This is the condmask *development* branch of coloc.</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coloc.abf.html">coloc.abf</a></span>(<span class="dt">dataset1=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b1<span class="op">$</span>beta, <span class="dt">varbeta=</span>b1<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X1),<span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y1),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">dataset2=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b2<span class="op">$</span>beta, <span class="dt">varbeta=</span>b2<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X2),<span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y2),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">MAF=</span>maf,<span class="dt">p12=</span><span class="fl">1e-6</span>)</code></pre></div>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  1.41e-36  2.59e-24  5.47e-15  2.41e-10  1.00e+00 
## [1] "PP abf for shared variant: 100%"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res</code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-06</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##         H0    H1    H2       H3    H4
##  0.9899255 0.005 0.005 2.45e-05 5e-05</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##        nsnps           H0           H1           H2           H3           H4 
## 5.000000e+01 1.413084e-36 2.585470e-24 5.465483e-15 2.410934e-10 1.000000e+00</code></pre>
<p>A sensitivity analysis can be used, post-hoc, to determine the range of prior probabilities for which a conclusion is still supported. The sensitivity() function shows this for variable <span class="math inline">\(p_{12}\)</span> in the bottom right plot, along with the prior probabilities of each hypothesis, which may help decide whether a particular range of <span class="math inline">\(p_{12}\)</span> is valid. The green region shows the region - the set of values of <span class="math inline">\(p_{12}\)</span> - for which <span class="math inline">\(H_4 &gt; 0.5\)</span> - the rule that was specified. In this case, the conclusion of colocalisation looks quite robust. On the left (optionally) the input data are also presented, with shading to indicate the posterior probabilities that a SNP is causal if <span class="math inline">\(H_4\)</span> were true. This can be useful to indicate serious discrepancies also.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(my.res,<span class="dt">rule=</span><span class="st">"H4 &gt; 0.5"</span>) </code></pre></div>
<pre><code>## Results pass decision rule H4 &gt; 0.5</code></pre>
<p><img src="a04_sensitivity_files/figure-html/sens-1.png" width="768"></p>
<p>Let’s make a smaller dataset where that won’t be the case:</p>
<p>Now, colocalisation is very dependent on the value of <span class="math inline">\(p_{12}\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res &lt;-<span class="st"> </span><span class="kw"><a href="../reference/coloc.abf.html">coloc.abf</a></span>(<span class="dt">dataset1=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b1<span class="op">$</span>beta, <span class="dt">varbeta=</span>b1<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X1),<span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y1),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">dataset2=</span><span class="kw"><a href="https://rdrr.io/r/base/list.html">list</a></span>(<span class="dt">beta=</span>b2<span class="op">$</span>beta, <span class="dt">varbeta=</span>b2<span class="op">$</span>varbeta, <span class="dt">N=</span><span class="kw"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span>(X2),<span class="dt">sdY=</span><span class="kw"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span>(Y2),<span class="dt">type=</span><span class="st">"quant"</span>),
                    <span class="dt">MAF=</span>maf,<span class="dt">p12=</span><span class="fl">1e-6</span>)</code></pre></div>
<pre><code>## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf 
##  0.005010  0.000444  0.122000  0.002100  0.871000 
## [1] "PP abf for shared variant: 87.1%"</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">my.res</code></pre></div>
<pre><code>## Coloc analysis of trait 1, trait 2</code></pre>
<pre><code>## 
## SNP Priors</code></pre>
<pre><code>##    p1    p2   p12 
## 1e-04 1e-04 1e-06</code></pre>
<pre><code>## 
## Hypothesis Priors</code></pre>
<pre><code>##         H0    H1    H2       H3    H4
##  0.9899255 0.005 0.005 2.45e-05 5e-05</code></pre>
<pre><code>## 
## Posterior</code></pre>
<pre><code>##        nsnps           H0           H1           H2           H3           H4 
## 5.000000e+01 5.010313e-03 4.440917e-04 1.219223e-01 2.101426e-03 8.705219e-01</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(my.res,<span class="dt">rule=</span><span class="st">"H4 &gt; 0.5"</span>) </code></pre></div>
<pre><code>## Results pass decision rule H4 &gt; 0.5</code></pre>
<p><img src="a04_sensitivity_files/figure-html/sens2-1.png" width="768"></p>
<p>In this case, we find there is evidence for colocalisation according to a rule <span class="math inline">\(H_4&gt;0.5\)</span> only for <span class="math inline">\(p_{12} &gt;&gt; 10^{-6}\)</span>, which corresponds to an <em>a priori</em> belief that <span class="math inline">\(P(H_4) &gt;&gt; P(H_3)\)</span>. This means but you would need to think it reasonable that <span class="math inline">\(H_4\)</span> is much more likely than <span class="math inline">\(H_3\)</span> to begin with to find these data convincing.</p>
<p>Note, the syntax can also consider more complicated rules:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/sensitivity.html">sensitivity</a></span>(my.res,<span class="dt">rule=</span><span class="st">"H4 &gt; 3*H3 &amp;&amp; H0 &lt; 0.1"</span>) </code></pre></div>
<pre><code>## Results pass decision rule H4 &gt; 3*H3 &amp;&amp; H0 &lt; 0.1</code></pre>
<p><img src="a04_sensitivity_files/figure-html/sens3-1.png" width="768"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#simulate-a-small-dataset">Simulate a small dataset</a></li>
      <li><a href="#sensitivity-analysis">Sensitivity analysis</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
