<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Coloc: sensitivity to prior values • coloc</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Coloc: sensitivity to prior values">
<meta property="og:description" content="coloc">
<meta property="og:image" content="https://chr1swallace.github.io/coloc/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a>
    </li>
    <li>
      <a href="../articles/a02_data.html">Coloc: data structures</a>
    </li>
    <li>
      <a href="../articles/a03_enumeration.html">Coloc: under a single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a>
    </li>
    <li>
      <a href="../articles/a06_SuSiE.html">Coloc: using SuSiE to relax the single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a05_conditioning.html">DEPRECATED Coloc: relaxing the single causal variant assumption</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/chr1swallace/coloc/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Coloc: sensitivity to prior values</h1>
                        <h4 data-toc-skip class="author">Chris
Wallace</h4>
            
            <h4 data-toc-skip class="date">2023-05-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/HEAD/vignettes/a04_sensitivity.Rmd" class="external-link"><code>vignettes/a04_sensitivity.Rmd</code></a></small>
      <div class="hidden name"><code>a04_sensitivity.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="sensitivity-analysis">Sensitivity analysis<a class="anchor" aria-label="anchor" href="#sensitivity-analysis"></a>
</h2>
<p>Specifying prior values for coloc.abf() is important, as results can
be dependent on these values. Defaults of <span class="math inline">\(p_1=p_2=10^{-4}\)</span> seem justified in a wide
range of scenarios, because these broadly correspond to a 99% belief
that there is true association when we see <span class="math inline">\(p&lt;5\times 10^{-8}\)</span> in a GWAS. However,
choice of <span class="math inline">\(p_{12}\)</span> is more difficult.
We hope the <a href="https://chr1swallace.shinyapps.io/coloc-priors/" class="external-link">coloc explorer
app</a> will be helpful in exploring what various choices mean, at a
per-SNP and per-hypothesis level. However, having conducted an
enumeration-based coloc analysis, it is still helpful to check that any
inference about colocalisation is robust to variations in prior values
specified.</p>
<p>Continuing on from <a href="./a03_enumeration.html">the last
vignette</a>, we have</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chr1swallace/coloc" class="external-link">coloc</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## This is coloc version 5.2.2</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">coloc_test_data</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/attach.html" class="external-link">attach</a></span><span class="op">(</span><span class="va">coloc_test_data</span><span class="op">)</span></span>
<span><span class="va">my.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coloc.abf.html">coloc.abf</a></span><span class="op">(</span>dataset1<span class="op">=</span><span class="va">D1</span>,</span>
<span>                    dataset2<span class="op">=</span><span class="va">D2</span>,</span>
<span>                    p12<span class="op">=</span><span class="fl">1e-6</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  1.37e-17  2.92e-09  8.53e-11  8.28e-03  9.92e-01 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 99.2%"</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my.res</span></span></code></pre></div>
<pre><code><span><span class="co">## Coloc analysis of trait 1, trait 2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SNP Priors</span></span></code></pre>
<pre><code><span><span class="co">##    p1    p2   p12 </span></span>
<span><span class="co">## 1e-04 1e-04 1e-06</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Hypothesis Priors</span></span></code></pre>
<pre><code><span><span class="co">##        H0   H1   H2       H3    H4</span></span>
<span><span class="co">##  0.897005 0.05 0.05 0.002495 5e-04</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Posterior</span></span></code></pre>
<pre><code><span><span class="co">##        nsnps           H0           H1           H2           H3           H4 </span></span>
<span><span class="co">## 5.000000e+02 1.366742e-17 2.915456e-09 8.529214e-11 8.276799e-03 9.917232e-01</span></span></code></pre>
<p>A sensitivity analysis can be used, post-hoc, to determine the range
of prior probabilities for which a conclusion is still supported. The
sensitivity() function shows this for variable <span class="math inline">\(p_{12}\)</span> in the bottom right plot, along
with the prior probabilities of each hypothesis, which may help decide
whether a particular range of <span class="math inline">\(p_{12}\)</span> is valid. The green region shows
the region - the set of values of <span class="math inline">\(p_{12}\)</span> - for which <span class="math inline">\(H_4 &gt; 0.5\)</span> - the rule that was
specified. In this case, the conclusion of colocalisation looks quite
robust. On the left (optionally) the input data are also presented, with
shading to indicate the posterior probabilities that a SNP is causal if
<span class="math inline">\(H_4\)</span> were true. This can be useful
to indicate serious discrepancies also.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sensitivity.html">sensitivity</a></span><span class="op">(</span><span class="va">my.res</span>,rule<span class="op">=</span><span class="st">"H4 &gt; 0.5"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Results pass decision rule H4 &gt; 0.5</span></span></code></pre>
<p><img src="a04_sensitivity_files/figure-html/sens-1.png" width="768"></p>
<p>Let’s fake a smaller dataset where that won’t be the case, by
increasing varbeta:</p>
<p>Now, colocalisation is very dependent on the value of <span class="math inline">\(p_{12}\)</span>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coloc.abf.html">coloc.abf</a></span><span class="op">(</span>dataset1<span class="op">=</span><span class="va">D1a</span>,</span>
<span>                    dataset2<span class="op">=</span><span class="va">D2a</span>,</span>
<span>                    p12<span class="op">=</span><span class="fl">1e-6</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  5.58e-07  1.61e-05  1.26e-03  2.67e-02  9.72e-01 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 97.2%"</span></span></code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my.res</span></span></code></pre></div>
<pre><code><span><span class="co">## Coloc analysis of trait 1, trait 2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SNP Priors</span></span></code></pre>
<pre><code><span><span class="co">##    p1    p2   p12 </span></span>
<span><span class="co">## 1e-04 1e-04 1e-06</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Hypothesis Priors</span></span></code></pre>
<pre><code><span><span class="co">##        H0   H1   H2       H3    H4</span></span>
<span><span class="co">##  0.897005 0.05 0.05 0.002495 5e-04</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Posterior</span></span></code></pre>
<pre><code><span><span class="co">##        nsnps           H0           H1           H2           H3           H4 </span></span>
<span><span class="co">## 5.000000e+02 5.583635e-07 1.612179e-05 1.261187e-03 2.669434e-02 9.720278e-01</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sensitivity.html">sensitivity</a></span><span class="op">(</span><span class="va">my.res</span>,rule<span class="op">=</span><span class="st">"H4 &gt; 0.5"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Results pass decision rule H4 &gt; 0.5</span></span></code></pre>
<p><img src="a04_sensitivity_files/figure-html/sens2-1.png" width="768"></p>
<p>In this case, we find there is evidence for colocalisation according
to a rule <span class="math inline">\(H_4&gt;0.5\)</span> only for <span class="math inline">\(p_{12} &gt; 10^{-6}\)</span>, which corresponds to
an <em>a priori</em> belief that <span class="math inline">\(P(H_4)
\simeq P(H_3)\)</span>. This means but you would need to think it
reasonable that <span class="math inline">\(H_4\)</span> is equally
likely as <span class="math inline">\(H_3\)</span> to begin with to find
these data convincing.</p>
<p>Note, the syntax can also consider more complicated rules:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sensitivity.html">sensitivity</a></span><span class="op">(</span><span class="va">my.res</span>,rule<span class="op">=</span><span class="st">"H4 &gt; 3*H3 &amp; H0 &lt; 0.1"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## Results pass decision rule H4 &gt; 3*H3 &amp; H0 &lt; 0.1</span></span></code></pre>
<p><img src="a04_sensitivity_files/figure-html/sens3-1.png" width="768"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
