<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>DEPRECATED Coloc: relaxing the single causal variant assumption • coloc</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/simplex/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="DEPRECATED Coloc: relaxing the single causal variant assumption">
<meta property="og:description" content="coloc">
<meta property="og:image" content="https://chr1swallace.github.io/coloc/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">coloc</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">5.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a>
    </li>
    <li>
      <a href="../articles/a02_data.html">Coloc: data structures</a>
    </li>
    <li>
      <a href="../articles/a03_enumeration.html">Coloc: under a single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a>
    </li>
    <li>
      <a href="../articles/a06_SuSiE.html">Coloc: using SuSiE to relax the single causal variant assumption</a>
    </li>
    <li>
      <a href="../articles/a05_conditioning.html">DEPRECATED Coloc: relaxing the single causal variant assumption</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/chr1swallace/coloc/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>DEPRECATED Coloc: relaxing the single causal
variant assumption</h1>
                        <h4 data-toc-skip class="author">Chris
Wallace</h4>
            
            <h4 data-toc-skip class="date">2023-05-12</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/HEAD/vignettes/a05_conditioning.Rmd" class="external-link"><code>vignettes/a05_conditioning.Rmd</code></a></small>
      <div class="hidden name"><code>a05_conditioning.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="multiple-causal-variants">Multiple causal variants<a class="anchor" aria-label="anchor" href="#multiple-causal-variants"></a>
</h2>
<p><em>This describes deprecated functions - the <a href="a06_SuSiE.html">SuSiE approach</a> is more accurate and should be
used instead</em></p>
<p>We load some simulated data.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chr1swallace/coloc" class="external-link">coloc</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## This is coloc version 5.2.2</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">coloc_test_data</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/attach.html" class="external-link">attach</a></span><span class="op">(</span><span class="va">coloc_test_data</span><span class="op">)</span> <span class="co"># contains D3, D4 that we will use in this vignette</span></span></code></pre></div>
<p>First, let us do a standard coloc (single causal variant) analysis to
serve as a baseline comparison. The analysis concludes there is
colocalisation, because it “sees” the SNPs on the left which are
strongly associated with both traits. But it misses the SNPs on the
right of the top left plot which are associated with only one trait.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chr1swallace/coloc" class="external-link">coloc</a></span><span class="op">)</span></span>
<span><span class="va">my.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coloc.abf.html">coloc.abf</a></span><span class="op">(</span>dataset1<span class="op">=</span><span class="va">D3</span>, dataset2<span class="op">=</span><span class="va">D4</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  8.78e-26  6.80e-07  1.53e-22  1.85e-04  1.00e+00 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 100%"</span></span></code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">my.res</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "coloc_abf" "list"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## print.coloc_abf</span></span>
<span><span class="va">my.res</span></span></code></pre></div>
<pre><code><span><span class="co">## Coloc analysis of trait 1, trait 2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SNP Priors</span></span></code></pre>
<pre><code><span><span class="co">##    p1    p2   p12 </span></span>
<span><span class="co">## 1e-04 1e-04 1e-05</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Hypothesis Priors</span></span></code></pre>
<pre><code><span><span class="co">##        H0   H1   H2       H3    H4</span></span>
<span><span class="co">##  0.892505 0.05 0.05 0.002495 0.005</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Posterior</span></span></code></pre>
<pre><code><span><span class="co">##        nsnps           H0           H1           H2           H3           H4 </span></span>
<span><span class="co">## 5.000000e+02 8.775708e-26 6.797736e-07 1.529399e-22 1.848705e-04 9.998144e-01</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sensitivity.html">sensitivity</a></span><span class="op">(</span><span class="va">my.res</span>,<span class="st">"H4 &gt; 0.9"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Results pass decision rule H4 &gt; 0.9</span></span></code></pre>
<p><img src="a05_conditioning_files/figure-html/sens0-1.png" width="768"></p>
<p>Even though the sensitivity analysis itself looks good, the Manhattan
plots suggest we are violating the assumption of a single causal variant
per trait.<br>
We can use =finemap.signals= to test whether there are additional
signals after conditioning.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/finemap.signals.html">finemap.signals</a></span><span class="op">(</span><span class="va">D3</span>,method<span class="op">=</span><span class="st">"cond"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      s105       s78 </span></span>
<span><span class="co">## 11.180489  5.351394</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/finemap.signals.html">finemap.signals</a></span><span class="op">(</span><span class="va">D4</span>,method<span class="op">=</span><span class="st">"cond"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    s105 </span></span>
<span><span class="co">## 6.42341</span></span></code></pre>
<p>Note that every colocalisation conditions out every other signal
except one for each trait. For that reason, trying to colocalise many
signals per trait is not recommended. Instead, use pthr to set the
significance (p value) required to call a signal. If you set if too low,
you will capture signals that are non-significant, or too high and you
will miss true signals. pthr=5e-8 would correspond to a genome-wide
significance level for common variants in a European study, but we
typically choose a slightly relaxed pthr=1e-6 on the basis that if there
is one GW-significant signal in a region, we expect there is a greater
chance for secondary signals to exist.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/finemap.signals.html">finemap.signals</a></span><span class="op">(</span><span class="va">D3</span>,method<span class="op">=</span><span class="st">"cond"</span>,pthr<span class="op">=</span><span class="fl">1e-20</span><span class="op">)</span> <span class="co">## too small</span></span></code></pre></div>
<pre><code><span><span class="co">##     s105 </span></span>
<span><span class="co">## 11.18049</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/finemap.signals.html">finemap.signals</a></span><span class="op">(</span><span class="va">D4</span>,method<span class="op">=</span><span class="st">"cond"</span>,pthr<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span> <span class="co">## too big</span></span></code></pre></div>
<pre><code><span><span class="co">##      s105      s108      s156 </span></span>
<span><span class="co">##  6.423410 -3.226812  3.741284</span></span></code></pre>
<p>Now we can ask coloc to consider these as separate signals using the
coloc.signals() function.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/coloc.signals.html">coloc.signals</a></span><span class="op">(</span><span class="va">D3</span>,<span class="va">D4</span>,method<span class="op">=</span><span class="st">"cond"</span>,p12<span class="op">=</span><span class="fl">1e-6</span>,pthr<span class="op">=</span><span class="fl">1e-6</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  8.76e-25  6.79e-06  1.53e-21  1.85e-03  9.98e-01 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 99.8%"</span></span>
<span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  1.83e-05  5.53e-04  3.19e-02  9.64e-01  3.52e-03 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 0.352%"</span></span>
<span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  5.44e-04  2.33e-05  9.49e-01  4.04e-02  1.02e-02 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 1.02%"</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span></span></code></pre></div>
<pre><code><span><span class="co">## Coloc analysis of trait 1, trait 2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SNP Priors</span></span></code></pre>
<pre><code><span><span class="co">##    p1    p2   p12 </span></span>
<span><span class="co">## 1e-04 1e-04 1e-06</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Hypothesis Priors</span></span></code></pre>
<pre><code><span><span class="co">##        H0   H1   H2       H3    H4</span></span>
<span><span class="co">##  0.897005 0.05 0.05 0.002495 5e-04</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Posterior</span></span></code></pre>
<pre><code><span><span class="co">##    nsnps hit1 hit2           H0           H1           H2         H3</span></span>
<span><span class="co">## 1:   500 s105 s105 8.674493e-25 6.719334e-06 1.511759e-21 0.01171021</span></span>
<span><span class="co">## 2:   500  s78 s105 1.830995e-05 5.531424e-04 3.190992e-02 0.96399665</span></span>
<span><span class="co">##             H4</span></span>
<span><span class="co">## 1: 0.988283067</span></span>
<span><span class="co">## 2: 0.003521974</span></span></code></pre>
<p>Note that because we are doing multiple colocalisations,
sensitivity() needs to know which to consider:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sensitivity.html">sensitivity</a></span><span class="op">(</span><span class="va">res</span>,<span class="st">"H4 &gt; 0.9"</span>,row<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Results pass decision rule H4 &gt; 0.9</span></span></code></pre>
<p><img src="a05_conditioning_files/figure-html/sens-1.png" width="768"></p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/sensitivity.html">sensitivity</a></span><span class="op">(</span><span class="va">res</span>,<span class="st">"H4 &gt; 0.9"</span>,row<span class="op">=</span><span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Results fail decision rule H4 &gt; 0.9</span></span></code></pre>
<p><img src="a05_conditioning_files/figure-html/sens-2.png" width="768"></p>
<!-- Because these signals are truly independent, we could also split them by using masking, which doesn't condition at all, but restricts the search space for colocalisation to SNPs which are not in LD with any-but-one of each signal SNP.  Here you also need to consider r2thr which sets the maximum \(r^2\) between two SNPs for them to be considered independent. -->
<!-- ```{r , fig.width=8,fig.height=6   } -->
<!-- finemap.signals(D3,method="mask",r2thr=0.3) -->
<!-- finemap.signals(D4,method="mask",r2thr=0.3) -->
<!-- resm=coloc.signals(D3,D4,method="mask",p12=1e-6,pthr=1e-6,r2thr=0.01) -->
<!-- resm -->
<!-- sensitivity(resm,"H4 > 0.9",row=1) -->
<!-- sensitivity(resm,"H4 > 0.9",row=2) -->
<!-- ``` -->
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Chris Wallace, Claudia Giambartolomei.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
