<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Coloc: using variant-specific priros in coloc • coloc</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Coloc: using variant-specific priros in coloc">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">coloc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">6.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a></li>
    <li><a class="dropdown-item" href="../articles/a02_data.html">Coloc: data structures</a></li>
    <li><a class="dropdown-item" href="../articles/a03_enumeration.html">Coloc: under a single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a></li>
    <li><a class="dropdown-item" href="../articles/a06_SuSiE.html">Coloc: using SuSiE to relax the single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a07_variant-specific-priors.html">Coloc: using variant-specific priros in coloc</a></li>
    <li><a class="dropdown-item" href="../articles/a05_conditioning.html">DEPRECATED Coloc: relaxing the single causal variant assumption</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chr1swallace/coloc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Coloc: using variant-specific priros in coloc</h1>
                        <h4 data-toc-skip class="author">Jeffrey
Pullin</h4>
            
            <h4 data-toc-skip class="date">2025-12-04</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/HEAD/vignettes/a07_variant-specific-priors.Rmd" class="external-link"><code>vignettes/a07_variant-specific-priors.Rmd</code></a></small>
      <div class="d-none name"><code>a07_variant-specific-priors.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="variant-specific-priors-in-coloc">Variant-specific priors in coloc<a class="anchor" aria-label="anchor" href="#variant-specific-priors-in-coloc"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chr1swallace/coloc" class="external-link">coloc</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## This is coloc version 6.0.1</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">coloc_test_data</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">eqtlgen_density_data</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/attach.html" class="external-link">attach</a></span><span class="op">(</span><span class="va">coloc_test_data</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/attach.html" class="external-link">attach</a></span><span class="op">(</span><span class="va">eqtlgen_density_data</span><span class="op">)</span></span></code></pre></div>
<p>By default coloc assumes that all variants in a region are equally
likely to be causal for the traits of interest. However, there is
substantial evidence that functional annotations and other genetic
information can provide evidence about whether variants are causal.</p>
</div>
<div class="section level2">
<h2 id="using-general-variant-specific-information">Using general variant-specific information<a class="anchor" aria-label="anchor" href="#using-general-variant-specific-information"></a>
</h2>
<p>To provide variant-specific information to coloc use the
<code>prior_weights1</code> or <code>prior_weights2</code> arguments for
<code><a href="../reference/coloc.abf.html">coloc.abf()</a></code>, <code><a href="../reference/coloc.susie.html">coloc.susie()</a></code> or
<code>coloc.bf()</code>. The arguments specify non-negative weights that
encode the ‘weight’ (proportional to the probability of a variant being
causal). Specifying weights for either or both of the traits will
automatically change the prior probability of a variant being causal for
<em>both</em> traits.</p>
<p>For example, if we had prior information that the first 100 variants
were twice as likely to be causal for trait 1 we could encode this in
the following prior weights:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">100</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">400</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/coloc.abf.html">coloc.abf</a></span><span class="op">(</span><span class="va">D1</span>, <span class="va">D2</span>, prior_weights1 <span class="op">=</span> <span class="va">weights</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  8.26e-19  2.91e-10  5.16e-12  8.18e-04  9.99e-01 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 99.9%"</span></span></code></pre>
<pre><code><span><span class="co">## Coloc analysis of trait 1, trait 2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SNP Priors</span></span></code></pre>
<pre><code><span><span class="co">##    p1    p2   p12 </span></span>
<span><span class="co">## 1e-04 1e-04 1e-05</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Hypothesis Priors</span></span></code></pre>
<pre><code><span><span class="co">##        H0   H1   H2       H3    H4</span></span>
<span><span class="co">##  0.892505 0.05 0.05 0.002495 0.005</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Posterior</span></span></code></pre>
<pre><code><span><span class="co">##        nsnps           H0           H1           H2           H3           H4 </span></span>
<span><span class="co">## 5.000000e+02 8.262373e-19 2.911768e-10 5.156170e-12 8.179198e-04 9.991821e-01</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="an-example-of-variant-specific-information">An example of variant-specific information<a class="anchor" aria-label="anchor" href="#an-example-of-variant-specific-information"></a>
</h2>
<p>Pullin and Wallace (2024) describe the implementation of
variant-specific priors in coloc and compare various sources of prior
information. The best performing source of prior information was the
density of the distance between an expression quantative trait loci
(eQTL) and the gene’s transcription start site (TSS). These estimated
densities show that eQTLs often lie close to the TSS, suggesting that we
should expect variants that lie closer to the TSS to be more likely to
be causal. The density estimated from the eQTLGen dataset, see Pullin
and Wallace (2024) for details, is stored in the coloc dataset
<code>eqtlgen_density_data</code> and we can visualise the density,
where the middle of the interval is the TSS.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span></span>
<span>  <span class="va">eqtlgen_density_data</span><span class="op">$</span><span class="va">rel_dist</span>,</span>
<span>  <span class="va">eqtlgen_density_data</span><span class="op">$</span><span class="va">dens_value</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="a07_variant-specific-priors_files/figure-html/unnamed-chunk-3-1.png" width="700"></p>
<p>To illustrate the use of the TSS we augment coloc’s simulated data
with simulated genomic positions in a 1Mb window around the TSS of
PTPN22. Then using a simple function we can compute prior weights for
these positions relative to different specified TSSs.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">113371759</span><span class="op">:</span><span class="fl">114371759</span>, <span class="fl">500</span>, replace <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">D1</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="va">pos</span></span>
<span><span class="va">D2</span><span class="op">$</span><span class="va">position</span> <span class="op">&lt;-</span> <span class="va">pos</span></span>
<span></span>
<span><span class="va">compute_eqtl_tss_dist_weights</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">pos</span>, <span class="va">tss</span>, <span class="va">density_data</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">rel</span> <span class="op">&lt;-</span> <span class="va">pos</span> <span class="op">-</span> <span class="va">tss</span></span>
<span>  <span class="va">closest</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">pos</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">closest</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.min</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">density_data</span><span class="op">$</span><span class="va">rel_dist</span> <span class="op">-</span> <span class="va">rel</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span>  <span class="va">out</span> <span class="op">&lt;-</span> <span class="va">density_data</span><span class="op">$</span><span class="va">dens_value</span><span class="op">[</span><span class="va">closest</span><span class="op">]</span></span>
<span>  <span class="va">out</span> <span class="op">&lt;-</span> <span class="va">out</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span>  <span class="va">out</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Computing the weights and applying them shows little impact on the
probability of colocalisation because this simulated example has very
strong evidence for colocalisation. In real data, the impact will likely
be greater.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the weights relative to the PTPN22 TSS.</span></span>
<span><span class="va">w1</span> <span class="op">&lt;-</span> <span class="fu">compute_eqtl_tss_dist_weights</span><span class="op">(</span><span class="va">D1</span><span class="op">$</span><span class="va">position</span>, <span class="fl">113871759</span>, <span class="va">eqtlgen_density_data</span><span class="op">)</span> </span>
<span></span>
<span><span class="fu"><a href="../reference/coloc.abf.html">coloc.abf</a></span><span class="op">(</span><span class="va">D1</span>, <span class="va">D2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  1.38e-18  2.94e-10  8.59e-12  8.34e-04  9.99e-01 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 99.9%"</span></span></code></pre>
<pre><code><span><span class="co">## Coloc analysis of trait 1, trait 2</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## SNP Priors</span></span></code></pre>
<pre><code><span><span class="co">##    p1    p2   p12 </span></span>
<span><span class="co">## 1e-04 1e-04 1e-05</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Hypothesis Priors</span></span></code></pre>
<pre><code><span><span class="co">##        H0   H1   H2       H3    H4</span></span>
<span><span class="co">##  0.892505 0.05 0.05 0.002495 0.005</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Posterior</span></span></code></pre>
<pre><code><span><span class="co">##        nsnps           H0           H1           H2           H3           H4 </span></span>
<span><span class="co">## 5.000000e+02 1.377000e-18 2.937336e-10 8.593226e-12 8.338917e-04 9.991661e-01</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/coloc.abf.html">coloc.abf</a></span><span class="op">(</span><span class="va">D1</span>, <span class="va">D2</span>, prior_weights1 <span class="op">=</span> <span class="va">w1</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## PP.H0.abf PP.H1.abf PP.H2.abf PP.H3.abf PP.H4.abf </span></span>
<span><span class="co">##  4.61e-17  1.19e-09  2.88e-10  6.44e-03  9.94e-01 </span></span>
<span><span class="co">## [1] "PP abf for shared variant: 99.4%"</span></span></code></pre>
<pre><code><span><span class="co">## Coloc analysis of trait 1, trait 2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## SNP Priors</span></span></code></pre>
<pre><code><span><span class="co">##    p1    p2   p12 </span></span>
<span><span class="co">## 1e-04 1e-04 1e-05</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Hypothesis Priors</span></span></code></pre>
<pre><code><span><span class="co">##        H0   H1   H2       H3    H4</span></span>
<span><span class="co">##  0.892505 0.05 0.05 0.002495 0.005</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Posterior</span></span></code></pre>
<pre><code><span><span class="co">##        nsnps           H0           H1           H2           H3           H4 </span></span>
<span><span class="co">## 5.000000e+02 4.607613e-17 1.191318e-09 2.875401e-10 6.440915e-03 9.935591e-01</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Wallace, Claudia Giambartolomei, Tom Willis, Jeffrey Pullin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
