<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Coloc with multiple signals per trait — coloc.signals • coloc</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Coloc with multiple signals per trait — coloc.signals"><meta name="description" content="New coloc function, builds on coloc.abf() by allowing for multiple
independent causal variants per trait through conditioning or
masking."><meta property="og:description" content="New coloc function, builds on coloc.abf() by allowing for multiple
independent causal variants per trait through conditioning or
masking."><meta property="og:image" content="https://chr1swallace.github.io/coloc/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">coloc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">6.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a></li>
    <li><a class="dropdown-item" href="../articles/a02_data.html">Coloc: data structures</a></li>
    <li><a class="dropdown-item" href="../articles/a03_enumeration.html">Coloc: under a single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a></li>
    <li><a class="dropdown-item" href="../articles/a06_SuSiE.html">Coloc: using SuSiE to relax the single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a07_variant-specific-priors.html">Coloc: using variant-specific priros in coloc</a></li>
    <li><a class="dropdown-item" href="../articles/a05_conditioning.html">DEPRECATED Coloc: relaxing the single causal variant assumption</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chr1swallace/coloc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Coloc with multiple signals per trait</h1>
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/HEAD/R/split.R" class="external-link"><code>R/split.R</code></a></small>
      <div class="d-none name"><code>coloc.signals.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>New coloc function, builds on coloc.abf() by allowing for multiple
independent causal variants per trait through conditioning or
masking.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">coloc.signals</span><span class="op">(</span></span>
<span>  <span class="va">dataset1</span>,</span>
<span>  <span class="va">dataset2</span>,</span>
<span>  MAF <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  LD <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"single"</span>, <span class="st">"cond"</span>, <span class="st">"mask"</span><span class="op">)</span>,</span>
<span>  mode <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"iterative"</span>, <span class="st">"allbutone"</span><span class="op">)</span>,</span>
<span>  p1 <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  p2 <span class="op">=</span> <span class="fl">1e-04</span>,</span>
<span>  p12 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  maxhits <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  r2thr <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  pthr <span class="op">=</span> <span class="fl">1e-06</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-dataset-">dataset1<a class="anchor" aria-label="anchor" href="#arg-dataset-"></a></dt>
<dd><p>a list with specifically named elements defining
the dataset to be analysed. See <code><a href="check_dataset.html">check_dataset</a></code>
for details.</p></dd>


<dt id="arg-dataset-">dataset2<a class="anchor" aria-label="anchor" href="#arg-dataset-"></a></dt>
<dd><p>as above, for dataset 2</p></dd>


<dt id="arg-maf">MAF<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>Common minor allele frequency vector to be used for
both dataset1 and dataset2, a shorthand for supplying the same
vector as parts of both datasets</p></dd>


<dt id="arg-ld">LD<a class="anchor" aria-label="anchor" href="#arg-ld"></a></dt>
<dd><p>required if method="cond". matrix of genotype
<em>correlation</em> (ie r, not r^2) between SNPs. If dataset1 and
dataset2 may have different LD, you can instead add LD=LD1 to
the list of dataset1 and a different LD matrix for dataset2</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>default "" means do no conditioning, should return
similar to coloc.abf.  if method="cond", then use conditioning
to coloc multiple signals.  if method="mask", use masking to
coloc multiple signals. if different datasets need different
methods (eg LD is only available for one of them) you can set
method on a per-dataset basis by adding method="..." to the
list for that dataset.</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>"iterative" or "allbutone".  Easiest understood with
an example.  Suppose there are 3 signal SNPs detected for
trait 1, A, B, C and only one for trait 2, D.</p>
<p></p><div class="sourceCode"><pre><code><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a>Under <span class="st">"iterative"</span> mode, <span class="dv">3</span> coloc will be performed<span class="sc">:</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="er">*</span> trait <span class="dv">1</span> <span class="sc">-</span> trait <span class="dv">2</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="sc">*</span> trait <span class="dv">1</span> conditioned on A <span class="sc">-</span> trait <span class="dv">2</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="sc">*</span> trait <span class="dv">1</span> conditioned on A<span class="sc">+</span>B <span class="sc">-</span> trait <span class="dv">2</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>Under <span class="st">"allbutone"</span> mode, they would be</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="sc">*</span> trait <span class="dv">1</span> conditioned on B<span class="sc">+</span>C <span class="sc">-</span> trait <span class="dv">2</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="sc">*</span> trait <span class="dv">1</span> conditioned on A<span class="sc">+</span>C <span class="sc">-</span> trait <span class="dv">2</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="sc">*</span> trait <span class="dv">1</span> conditioned on A<span class="sc">+</span>B <span class="sc">-</span> trait <span class="dv">2</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>Only iterative mode is supported <span class="cf">for</span> method<span class="ot">=</span><span class="st">"mask"</span>.</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a>The allbutone mode is optimal <span class="cf">if</span> the signals are known with</span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="fu">certainty</span> (which they never are), because it allows each</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>signal to be tested without influence of the others.  When</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>there is uncertainty, it may make sense to use iterative mode,</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>because the strongest signals aren<span class="st">'t affected by conditioning</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="st">incorrectly on weaker secondary and less certain signals.</span></span></code></pre><p></p></div></dd>


<dt id="arg-p-">p1<a class="anchor" aria-label="anchor" href="#arg-p-"></a></dt>
<dd><p>prior probability a SNP is associated with trait 1,
default 1e-4</p></dd>


<dt id="arg-p-">p2<a class="anchor" aria-label="anchor" href="#arg-p-"></a></dt>
<dd><p>prior probability a SNP is associated with trait 2,
default 1e-4</p></dd>


<dt id="arg-p-">p12<a class="anchor" aria-label="anchor" href="#arg-p-"></a></dt>
<dd><p>prior probability a SNP is associated with both traits,
default 1e-5</p></dd>


<dt id="arg-maxhits">maxhits<a class="anchor" aria-label="anchor" href="#arg-maxhits"></a></dt>
<dd><p>maximum number of levels to condition/mask</p></dd>


<dt id="arg-r-thr">r2thr<a class="anchor" aria-label="anchor" href="#arg-r-thr"></a></dt>
<dd><p>if masking, the threshold on r2 should be used to
call two signals independent.  our experience is that this
needs to be set low to avoid double calling the same strong
signal.</p></dd>


<dt id="arg-pthr">pthr<a class="anchor" aria-label="anchor" href="#arg-pthr"></a></dt>
<dd><p>if masking or conditioning, what p value threshold to
call a secondary hit "significant"</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>data.table of coloc results, one row per pair of lead snps
detected in each dataset</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Chris Wallace</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Wallace, Claudia Giambartolomei, Tom Willis, Jeffrey Pullin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

