<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ coloc</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Package index"><meta property="og:image" content="https://chr1swallace.github.io/coloc/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">coloc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">6.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a></li>
    <li><a class="dropdown-item" href="../articles/a02_data.html">Coloc: data structures</a></li>
    <li><a class="dropdown-item" href="../articles/a03_enumeration.html">Coloc: under a single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a></li>
    <li><a class="dropdown-item" href="../articles/a06_SuSiE.html">Coloc: using SuSiE to relax the single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a07_variant-specific-priors.html">Coloc: using variant-specific priros in coloc</a></li>
    <li><a class="dropdown-item" href="../articles/a05_conditioning.html">DEPRECATED Coloc: relaxing the single causal variant assumption</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chr1swallace/coloc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="coloc-data-structure">Coloc data structure<a class="anchor" aria-label="anchor" href="#coloc-data-structure"></a></h2>

      <div class="section-desc"><p>Description of coloc data structure</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="check_dataset.html">check_dataset()</a></code> <code><a href="check_dataset.html">check.dataset()</a></code>

        </dt>
        <dd>check_dataset</dd>
      </dl></div><div class="section level2">
      <h2 id="bayesian-enumeration-of-hypotheses">Bayesian enumeration of hypotheses<a class="anchor" aria-label="anchor" href="#bayesian-enumeration-of-hypotheses"></a></h2>

      <div class="section-desc"><p>Two trait colocalisation and single trait fine mapping under single causal variant assumptions</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="coloc.abf.html">coloc.abf()</a></code>

        </dt>
        <dd>Fully Bayesian colocalisation analysis using Bayes Factors</dd>
      </dl><dl><dt>

          <code><a href="finemap.abf.html">finemap.abf()</a></code>

        </dt>
        <dd>Bayesian finemapping analysis</dd>
      </dl></div><div class="section level2">
      <h2 id="sensitivity-analysis">Sensitivity analysis<a class="anchor" aria-label="anchor" href="#sensitivity-analysis"></a></h2>

      <div class="section-desc"><p>Check how robust inference is to changing prior parameter values</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="sensitivity.html">sensitivity()</a></code>

        </dt>
        <dd>Prior sensitivity for coloc</dd>
      </dl></div><div class="section level2">
      <h2 id="relaxing-the-single-variant-assumption">Relaxing the single variant assumption<a class="anchor" aria-label="anchor" href="#relaxing-the-single-variant-assumption"></a></h2>

      <div class="section-desc"><p>use conditioning or masking to account for multiple independent causal variants per trait</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="coloc.susie.html">coloc.susie()</a></code>

        </dt>
        <dd>run coloc using susie to detect separate signals</dd>
      </dl><dl><dt>

          <code><a href="runsusie.html">runsusie()</a></code>

        </dt>
        <dd>Run susie on a single coloc-structured dataset</dd>
      </dl></div><div class="section level2">
      <h2 id="visualisation">Visualisation<a class="anchor" aria-label="anchor" href="#visualisation"></a></h2>

      <div class="section-desc"><p>see stuff</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="plot_dataset.html">plot_dataset()</a></code>

        </dt>
        <dd>plot a coloc dataset</dd>
      </dl><dl><dt>

          <code><a href="plot_datasets.html">plot_datasets()</a></code>

        </dt>
        <dd>plot a pair of coloc datasets, highlighting the snps that are common between them</dd>
      </dl><dl><dt>

          <code><a href="plot_extended_datasets.html">plot_extended_datasets()</a></code>

        </dt>
        <dd>Draw extended plot of summary statistics for two coloc datasets</dd>
      </dl></div><div class="section level2">
      <h2 id="others">Others<a class="anchor" aria-label="anchor" href="#others"></a></h2>

      <div class="section-desc"><p>included for completeness, but the main functions you need should be in the sections above</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="annotate_susie.html">annotate_susie()</a></code>

        </dt>
        <dd>annotate susie_rss output for use with coloc_susie</dd>
      </dl><dl><dt>

          <code><a href="approx.bf.estimates.html">approx.bf.estimates()</a></code>

        </dt>
        <dd>Internal function, approx.bf.estimates</dd>
      </dl><dl><dt>

          <code><a href="approx.bf.p.html">approx.bf.p()</a></code>

        </dt>
        <dd>Internal function, approx.bf.p</dd>
      </dl><dl><dt>

          <code><a href="bin2lin.html">bin2lin()</a></code>

        </dt>
        <dd>binomial to linear regression conversion</dd>
      </dl><dl><dt>

          <code><a href="check_alignment.html">check_alignment()</a></code> <code><a href="check_alignment.html">check.alignment()</a></code>

        </dt>
        <dd>check alignment</dd>
      </dl><dl><dt>

          <code><a href="coloc-package.html">-package</a></code> <code><a href="coloc-package.html">coloc-package</a></code>

        </dt>
        <dd>Colocalisation tests of two genetic traits</dd>
      </dl><dl><dt>

          <code><a href="coloc.bf_bf.html">coloc.bf_bf()</a></code>

        </dt>
        <dd>Coloc data through Bayes factors</dd>
      </dl><dl><dt>

          <code><a href="coloc.detail.html">coloc.detail()</a></code>

        </dt>
        <dd>Bayesian colocalisation analysis with detailed output</dd>
      </dl><dl><dt>

          <code><a href="coloc.process.html">coloc.process()</a></code>

        </dt>
        <dd>Post process a coloc.details result using masking</dd>
      </dl><dl><dt>

          <code><a href="coloc.signals.html">coloc.signals()</a></code>

        </dt>
        <dd>Coloc with multiple signals per trait</dd>
      </dl><dl><dt>

          <code><a href="coloc.susie_bf.html">coloc.susie_bf()</a></code>

        </dt>
        <dd>run coloc using susie to detect separate signals</dd>
      </dl><dl><dt>

          <code><a href="coloc_test_data.html">coloc_test_data</a></code>

        </dt>
        <dd>Simulated data to use in testing and vignettes in the coloc package</dd>
      </dl><dl><dt>

          <code><a href="combine.abf.html">combine.abf()</a></code>

        </dt>
        <dd>combine.abf</dd>
      </dl><dl><dt>

          <code><a href="credible.sets.html">credible.sets()</a></code>

        </dt>
        <dd>credible.sets</dd>
      </dl><dl><dt>

          <code><a href="est_cond.html">est_cond()</a></code>

        </dt>
        <dd>generate conditional summary stats</dd>
      </dl><dl><dt>

          <code><a href="estgeno1.html">estgeno.1.ctl()</a></code> <code><a href="estgeno1.html">estgeno.1.cse()</a></code>

        </dt>
        <dd>estgeno1</dd>
      </dl><dl><dt>

          <code><a href="find.best.signal.html">find.best.signal()</a></code>

        </dt>
        <dd>Pick out snp with most extreme Z score</dd>
      </dl><dl><dt>

          <code><a href="findends.html">findends()</a></code>

        </dt>
        <dd>trim a dataset to central peak(s)</dd>
      </dl><dl><dt>

          <code><a href="findpeaks.html">findpeaks()</a></code>

        </dt>
        <dd>trim a dataset to only peak(s)</dd>
      </dl><dl><dt>

          <code><a href="finemap.bf.html">finemap.bf()</a></code>

        </dt>
        <dd>Finemap data through Bayes factors</dd>
      </dl><dl><dt>

          <code><a href="finemap.signals.html">finemap.signals()</a></code>

        </dt>
        <dd>Finemap multiple signals in a single dataset</dd>
      </dl><dl><dt>

          <code><a href="logbf_to_pp.html">logbf_to_pp()</a></code>

        </dt>
        <dd>logbf 2 pp</dd>
      </dl><dl><dt>

          <code><a href="logdiff.html">logdiff()</a></code>

        </dt>
        <dd>logdiff</dd>
      </dl><dl><dt>

          <code><a href="logsum.html">logsum()</a></code>

        </dt>
        <dd>logsum</dd>
      </dl><dl><dt>

          <code><a href="map_cond.html">map_cond()</a></code>

        </dt>
        <dd>find the next most significant SNP, conditioning on a list of sigsnps</dd>
      </dl><dl><dt>

          <code><a href="map_mask.html">map_mask()</a></code>

        </dt>
        <dd>find the next most significant SNP, masking a list of sigsnps</dd>
      </dl><dl><dt>

          <code><a href="plot-methods.html">plot(<i>&lt;coloc_abf&gt;</i>)</a></code>

        </dt>
        <dd>plot a coloc_abf object</dd>
      </dl><dl><dt>

          <code><a href="print-methods.html">print(<i>&lt;coloc_abf&gt;</i>)</a></code>

        </dt>
        <dd>print.coloc_abf</dd>
      </dl><dl><dt>

          <code><a href="process.dataset.html">process.dataset()</a></code>

        </dt>
        <dd>process.dataset</dd>
      </dl><dl><dt>

          <code><a href="sdY.est.html">sdY.est()</a></code>

        </dt>
        <dd>Estimate trait variance, internal function</dd>
      </dl><dl><dt>

          <code><a href="subset_dataset.html">subset_dataset()</a></code>

        </dt>
        <dd>subset_dataset</dd>
      </dl><dl><dt>

          <code><a href="Var.data.cc.html">Var.data.cc()</a></code>

        </dt>
        <dd>Var.data</dd>
      </dl><dl><dt>

          <code><a href="Var.data.html">Var.data()</a></code>

        </dt>
        <dd>Var.data</dd>
      </dl><dl><dt>

          <code><a href="eqtlgen_density_data.html">eqtlgen_density_data</a></code>

        </dt>
        <dd>eQTLGen estimated distance density</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Wallace, Claudia Giambartolomei, Tom Willis, Jeffrey Pullin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

