<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Draw extended plot of summary statistics for two coloc datasets — plot_extended_datasets • coloc</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Draw extended plot of summary statistics for two coloc datasets — plot_extended_datasets"><meta name="description" content="Draw Manhattan-style locus plots for p-values in each dataset, gene annotations,
a scatter plot of z-scores, and a table of coloc summary statistics."><meta property="og:description" content="Draw Manhattan-style locus plots for p-values in each dataset, gene annotations,
a scatter plot of z-scores, and a table of coloc summary statistics."><meta property="og:image" content="https://chr1swallace.github.io/coloc/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">coloc</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">6.0.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/a01_intro.html">Coloc: a package for colocalisation analyses</a></li>
    <li><a class="dropdown-item" href="../articles/a02_data.html">Coloc: data structures</a></li>
    <li><a class="dropdown-item" href="../articles/a03_enumeration.html">Coloc: under a single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a04_sensitivity.html">Coloc: sensitivity to prior values</a></li>
    <li><a class="dropdown-item" href="../articles/a06_SuSiE.html">Coloc: using SuSiE to relax the single causal variant assumption</a></li>
    <li><a class="dropdown-item" href="../articles/a07_variant-specific-priors.html">Coloc: using variant-specific priros in coloc</a></li>
    <li><a class="dropdown-item" href="../articles/a05_conditioning.html">DEPRECATED Coloc: relaxing the single causal variant assumption</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/chr1swallace/coloc/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Draw extended plot of summary statistics for two coloc datasets</h1>
      <small class="dont-index">Source: <a href="https://github.com/chr1swallace/coloc/blob/HEAD/R/plot.R" class="external-link"><code>R/plot.R</code></a></small>
      <div class="d-none name"><code>plot_extended_datasets.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Draw Manhattan-style locus plots for p-values in each dataset, gene annotations,
a scatter plot of z-scores, and a table of coloc summary statistics.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">plot_extended_datasets</span><span class="op">(</span></span>
<span>  <span class="va">d1</span>,</span>
<span>  <span class="va">d2</span>,</span>
<span>  <span class="va">x</span>,</span>
<span>  first_highlight_list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  second_highlight_list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  first_index_snp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  second_index_snp <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  first_trait <span class="op">=</span> <span class="st">"Trait 1"</span>,</span>
<span>  second_trait <span class="op">=</span> <span class="st">"Trait 2"</span>,</span>
<span>  snp_label <span class="op">=</span> <span class="st">"snp"</span>,</span>
<span>  ld_label <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  show_ld <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  locus_plot_ylim <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ens_db <span class="op">=</span> <span class="st">"EnsDb.Hsapiens.v86"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-d-">d1<a class="anchor" aria-label="anchor" href="#arg-d-"></a></dt>
<dd><p>a coloc dataset</p></dd>


<dt id="arg-d-">d2<a class="anchor" aria-label="anchor" href="#arg-d-"></a></dt>
<dd><p>another coloc dataset</p></dd>


<dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>object of class <code>coloc_abf</code> returned by <code><a href="coloc.abf.html">coloc.abf()</a></code></p></dd>


<dt id="arg-first-highlight-list">first_highlight_list<a class="anchor" aria-label="anchor" href="#arg-first-highlight-list"></a></dt>
<dd><p>character vector of snps to highlight in the first dataset;
<code>'index'</code> can be passed to highlight the most significant SNP</p></dd>


<dt id="arg-second-highlight-list">second_highlight_list<a class="anchor" aria-label="anchor" href="#arg-second-highlight-list"></a></dt>
<dd><p>character vector of snps to highlight in the second dataset;
<code>'index'</code> can be passed to highlight the most significant SNP</p></dd>


<dt id="arg-first-index-snp">first_index_snp<a class="anchor" aria-label="anchor" href="#arg-first-index-snp"></a></dt>
<dd><p>snp to designate as index snp in the first dataset
for the purpose of LD visualisation</p></dd>


<dt id="arg-second-index-snp">second_index_snp<a class="anchor" aria-label="anchor" href="#arg-second-index-snp"></a></dt>
<dd><p>snp to designate as index snp in the second dataset
for the purpose of LD visualisation</p></dd>


<dt id="arg-first-trait">first_trait<a class="anchor" aria-label="anchor" href="#arg-first-trait"></a></dt>
<dd><p>label for the first trait</p></dd>


<dt id="arg-second-trait">second_trait<a class="anchor" aria-label="anchor" href="#arg-second-trait"></a></dt>
<dd><p>label for the second trait</p></dd>


<dt id="arg-snp-label">snp_label<a class="anchor" aria-label="anchor" href="#arg-snp-label"></a></dt>
<dd><p>label for the snp column</p></dd>


<dt id="arg-ld-label">ld_label<a class="anchor" aria-label="anchor" href="#arg-ld-label"></a></dt>
<dd><p>label for the LD column</p></dd>


<dt id="arg-show-ld">show_ld<a class="anchor" aria-label="anchor" href="#arg-show-ld"></a></dt>
<dd><p>logical, whether to show LD in the locus plots</p></dd>


<dt id="arg-locus-plot-ylim">locus_plot_ylim<a class="anchor" aria-label="anchor" href="#arg-locus-plot-ylim"></a></dt>
<dd><p>numeric vector of length 2 specifying the y-axis
limits for the locus plots on the -log10 scale</p></dd>


<dt id="arg-ens-db">ens_db<a class="anchor" aria-label="anchor" href="#arg-ens-db"></a></dt>
<dd><p>character string specifying Ensembl database package from which to get gene positions. Current (at time of writing this documentation!) sensible values are  "EnsDb.Hsapiens.v86" for build 38 and "EnsDb.Hsapiens.v75" for build 37.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a gtable object</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Plot a pair of coloc datasets in an extended format</p>
<p>This function expects that the first two elements of the coloc
dataset list d contain summary statistics. Columns labelled 'chromosome',
'position', and 'p' (p-values) are expected in each dataset. The packages <code>locuszoomr</code>
and one of <code>EnsDb.Hsapiens.v86</code> or <code>EnsDb.Hsapiens.v75</code> are required
for this function. One of the <code>EnsDb.Hsapiens</code> libraries containing the Ensembl database
specified by ens_db must be loaded prior to  use (see the example). EnsDb.Hsapiens.v86
is the default database (GRCh38/hg19); for GRCh37/hg19, use EnsDb.Hsapiens.v75.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Tom Willis</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/chr1swallace/coloc" class="external-link">coloc</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu">plot_extended_dataset</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">first_dataset</span>, <span class="va">second_dataset</span><span class="op">)</span>, <span class="va">coloc</span>,</span></span>
<span class="r-in"><span>first_highlight_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs123"</span>, <span class="st">"rs456"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>second_highlight_list <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rs789"</span>, <span class="st">"rs1011"</span><span class="op">)</span>, </span></span>
<span class="r-in"><span>ens_db <span class="op">=</span> <span class="st">"EnsDb.Hsapiens.v86"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Chris Wallace, Claudia Giambartolomei, Tom Willis, Jeffrey Pullin.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

